<!doctype html>
<html lang="en">

<head>
  <!-- Required meta tags -->
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width">

  <!-- Bootstrap CSS -->
  <link rel="stylesheet" href="bootstrap.min.css" type="text/css">
  <link rel="stylesheet" href="course.css" type="text/css">

  <style>
    .jumbotron-head {
      background: linear-gradient(rgba(0, 0, 250, 0.15), rgba(125, 250, 250, 0.55)), url(img/Head_microbiome.jpg);
      background-repeat: no-repeat;
      background-attachment: fixed;
      background-position: top right;
      color: black !important;
    }
  </style>


  <title>Microbiome - Pathogen 'Omics</title>
</head>

<body>
  <div class="jumbotron jumbotron-fluid jumbotron-head">
    <div class="container">
      <h1 class="display-4">Microbiome</h1>
      <p class="lead">A case study: Bacterial Vaginosis</p>
    </div> <!-- Container -->
  </div>
  <div class="container">
    <h4>Microbiomes lecture</h4><br>
    <object data="presentations/microbiome_lecture.pdf#page=1" type="application/pdf" width="100%" height=700px>
      <p><b>Example fallback content</b>: This browser does not support PDFs. Please download the PDF to view it:
        <a href="presentations/microbiome_lecture.pdf">Download PDF</a>.</p>
    </object>
    <hr>
    <h4>Microbiomes Practical</h4><br>
    <p> Bacterial vaginosis (BV) is a dysbiotic condition caused by excessive growth of certain bacteria replacing the
      regular vaginal microbiome. Common symptoms include increased discharge, burning with urination, and itching. BV
      increases the risk of infection by a number of sexually transmitted infections including HIV/AIDS as well as the
      risk of early delivery when pregnant. The changed composition of the microbiome leads to a higher pH and a hundred
      to thousand-fold increase in the total number of bacteria present.
    </p>
    <p>For this practical we are considering 12 samples of vaginal swab that were taken at a polyclinic by a GP in a
      setting of high
      transmission of HIV. DNA was extracted from the swabs and amplified using primers specific for the first two
      hypervariable regions (V1 and V2) of the 16S
      rRNA gene (27F and 338R). These samples were then sequenced with MiSeq Illumina producing
      paired end data of 300&nbsp;bp length per read. The 12 pairs of files generated are found in the
      <code>data/metagenomics/fastq/</code> directory. The patients’
      phenotype was determined by the doctors at the time of sample collection with the following results:</p>


    <div class="row">
      <div class="col-sm-2"></div>
      <div class="col-sm-8">
        <table class="table">
          <thead class="thead-dark">
            <tr>
              <th style="text-align:center" , scope="col">Sample</th>
              <th style="text-align:center" , scope="col">BV</th>
              <th style="text-align:center" , scope="col">pH</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th style="text-align:center" , scope="row">BB_1</th>
              <td style="text-align:center">no</td>
              <td style="text-align:center">4.4</td>
            </tr>
            <tr>
              <th style="text-align:center" , scope="row">BB_2</th>
              <td style="text-align:center">no</td>
              <td style="text-align:center">3.6</td>
            </tr>
            <tr>
              <th style="text-align:center" , scope="row">BB_3</th>
              <td style="text-align:center">yes</td>
              <td style="text-align:center">5.5</td>
            </tr>
            <tr>
              <th style="text-align:center" , scope="row">BB_4</th>
              <td style="text-align:center">no</td>
              <td style="text-align:center">5.3</td>
            </tr>
            <tr>
              <th style="text-align:center" , scope="row">BB_5</th>
              <td style="text-align:center">yes</td>
              <td style="text-align:center">5.6</td>
            </tr>
            <tr>
              <th style="text-align:center" , scope="row">BB_6</th>
              <td style="text-align:center">yes</td>
              <td style="text-align:center">5.3</td>
            </tr>
            <tr>
              <th style="text-align:center" , scope="row">BB_7</th>
              <td style="text-align:center">yes</td>
              <td style="text-align:center">4.7</td>
            </tr>
            <tr>
              <th style="text-align:center" , scope="row">BB_8</th>
              <td style="text-align:center">no</td>
              <td style="text-align:center">4.4</td>
            </tr>
            <tr>
              <th style="text-align:center" , scope="row">BB_9</th>
              <td style="text-align:center">no</td>
              <td style="text-align:center">4.4</td>
            </tr>
            <tr>
              <th style="text-align:center" , scope="row">BB_10</th>
              <td style="text-align:center">no</td>
              <td style="text-align:center">3.6</td>
            </tr>
            <tr>
              <th style="text-align:center" , scope="row">BB_11</th>
              <td style="text-align:center">yes</td>
              <td style="text-align:center">4.7</td>
            </tr>
            <tr>
              <th style="text-align:center" , scope="row">BB_12</th>
              <td style="text-align:center">yes</td>
              <td style="text-align:center">5.0</td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>


    <h4>Analysing the microbiome samples with <i>QIIME2</i></h4>
    <p>Out of all tookits aiming at the unification of the analysis of microbiome data, QIIME (pronounced "chime") and
      its successor QIIME2 have grown the largest user base in recent years (mostly due to the ease of use
      and
      comprehensive <a href=https://docs.qiime2.org/>online documentation</a>). QIIME2 wraps an extensive suite of third
        party tools (covering most of the "standard" microbiome pipeline from preprocessing and filtering of raw
        sequencing reads to statistical tests on diversity metrics and analyses on differential abundance of single
        taxa) into a single command line interface. In addition, it also provides a GUI as well as a python API for both
        less and more technically inclined users. We will stay on the middle ground by using the CLI today. <br>
        One idiosyncrasy of QIIME2 is the
        use of
        so-called "artefacts". These are zip-archives with a special file extension (<code>.qza</code> for data
        artefacts and <code>.qzv</code> for visualisation artefacts) that hold bulk data in addition to unique IDs and
        provenance metadata, which describe all steps that lead to the creation of that particular artefact. This has
        the
        advantage that for every intermediate or final result of <code>qiime</code> it is perfectly clear how it was
        generated from start to finish. There is a small downside, though, since we have to import our data into the
        QIIME2 format prior to running any analyses. However, before we do that, let's have a look at the quality of
        our reads (<code>qiime</code> also provides functionality for sequencing data quality control, but it is not as
        detailed as the output of some dedicated tools like <a
          href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>).
    </p>
    <h5>Quality control</h5>
    <p>
      After activating the <code>conda</code> environment for this practical with <kbd>conda activate
        microbiome</kbd>, go into the module directory with <kbd>cd data/metagenomics</kbd> and have a look at its
      contents with
      <kbd>ls</kbd>. There should be a directory with the 16S sequencing data (in <code>fastq</code>), a 16S database
      (in <code>db</code>), and a CSV file with our metadata. Let's check if our reads are there with <kbd>ls
        fastq</kbd>. We can also have a look at the filesizes with <kbd>du -sh fastq/* | sort -h</kbd> (it can't hurt
      to get a feeling for these things).
      <br>
      Now that we have made sure that we have all the reads, let's run <code>fastqc</code> on them and store the
      reports in a new directory by running
      <pre>mkdir fastqc_reports</pre>
      <pre>fastqc -o fastqc_reports -q -t 1 fastq/*</pre>
      <div class="alert alert-primary">
        The <code>-t</code> flag tells <code>fastqc</code> the number of threads to use. If you have more CPUs
        available, adjust this number accordingly.
      </div>
      This should produce a FastQC report for each <code>fastq</code> file and put them all into
      <code>fastqc_reports</code> (run <kbd>ls fastqc_reports</kbd> to double check). Going through 24 FastQC reports
      (two per sample; one for the forward and one for the reverse reads) manually would be quite tedious. Thankfully,
      <code>multiqc</code> can combine them for us! Let's create a new directory for it to write the results into and
      run it.
      <pre>mkdir fastqc_combined</pre>
      <pre>multiqc -o fastqc_combined fastqc_reports</pre>
      With <kbd>ls fastqc_combined</kbd> you can see that an HTML file, which we can view in a browser, has been
      created.
      <div class="alert alert-primary">
        If running a VM or remote desktop, you should be able to simply double-click on the file in the file browser
        in order for it to open (the command line equivalent of double-clicking would be <kbd>xdg-open
          fastqc_combined/multiqc_report.html</kbd> in this case). If you are following the practical using a remote
        machine via
        SSH, however, using the browser on the server-side would be very slow. Instead, you would preferably either
        transfer
        the file to your local machine by typing <kbd>scp&nbsp;user@ipaddress:/absolute/path/on/remote&nbsp;.</kbd> into
        a local terminal, connect
        to the remote server in your local file browser if you
        are running macOS or Linux, or open it from the file panel in MobaXterm / PuTTY on Windows.
      </div>
      <div class="alert alert-primary">
        Hint: If you quickly need to get the absolute path of a file (e.g. for use in <code>scp</code>), type
        <kbd>find&nbsp;$PWD/filename</kbd> and copy the output.
      </div>
      Scroll through the report and make note of the sequence counts barplots and the quality histograms.
    </p>
    <div class="text-center"><img src="img/microbiome_1.png" class="img-fluid"></div>
    <div class="text-center"><img src="img/microbiome_2.png" class="img-fluid"></div>
    <p>
      In the
      counts plot we can see that two samples (BB_3 and BB_8) have significantly fewer reads than the others. This looks
      like something went wrong during library preparation for these two samples and we should exclude them
      from further analysis. Also note that the reads of the other samples have decent quality up until ~200&nbsp;bp
      length. We will need this information later.
      <div class="alert alert-warning">
        The difference in file sizes of the sequencing files between BB_3 / BB_8 and the other samples was not as
        drastic as the difference in actual read counts. Can you think of a reason why that might be? Hint: What does
        the file extension <code>.gz</code> mean?
      </div>
      For this dataset, the primers have already been trimmed from the reads and the FastQC output showed us that
      there are no adapters that would need removal. Also, quality-based trimming is <a
        href="https://github.com/benjjneb/dada2/issues/236#issuecomment-406606167">discouraged</a> when using <a
        href="https://www.nature.com/articles/nmeth.3869">DADA2</a> (an important step in our later analyses).
      Therefore, no
      further pre-processing is needed and we can transform the data into the <code>qiime</code> format.
    </p>

    <h5>Import into <code>qiime</code></h5>
    <p>
      In order to do this, <code>qiime</code> needs a tab-separated file with the sample
      IDs and the absolute paths to the forward and reverse reads. There are many ways to create such a file (and if
      you have just a few samples you can simply type it by hand). We will use the opportunity to string a few handy
      command line utilities together that we have not seen so far. First, let's write the header line of the
      import-list to a new file:
    </p>
    <pre><code>printf \
    "sample-id\tforward-absolute-filepath\treverse-absolute-filepath\n" \
    > fastq_abs_paths</code></pre>
    <div class="alert alert-primary" role="alert">
      We use backslashs here to break this command into multiple lines.
    </div>
    Then, we append the lines corresponding to our samples to the file that was just created. We can achieve this
    with
    <pre>ls fastq | grep -oE 'BB_[0-9]+' | sort -t _ -k 2 -n | uniq | \
      grep -vE 'BB_[38]' | \
      awk -v path=$(pwd)/fastq/ 'OFS="\t" \
          {print $1, path $1 "_1.fastq.gz", path $1 "_2.fastq.gz"}' \
      >> fastq_abs_paths</pre>
    <div class="alert alert-primary" role="alert">
      With the first <code>grep</code> we get the sample IDs from the filenames. We then <code>sort</code> them
      numerically (<code>-n</code>) based on the second field (<code>-k 2</code>) when split at underscores (<code>-t
        _</code>). Since there are two files per sample, we only keep the <code>uniq</code> sample IDs before removing
      the low-read-counts samples (BB_3 and BB_8) with another <code>grep</code> (<code>grep -v</code> keeps all lines
      that do
      <i>not</i> match the <a href="https://en.wikipedia.org/wiki/Regular_expression">regex</a>). The remaining sample
      IDs are subsequently fed to <code>awk</code> in order to print the absolute paths which are finally
      appended to <code>fastq_abs_paths</code>.
    </div>
    This should have done the trick. <kbd>cat fastq_abs_paths</kbd> let's us see what we got. The output should look
    like this:
    <pre>sample-id       forward-absolute-filepath       reverse-absolute-filepath
BB_1    /home/user/data/metagenomics/fastq/BB_1_1.fastq.gz      /home/user/data/metagenomics/fastq/BB_1_2.fastq.gz
BB_2    /home/user/data/metagenomics/fastq/BB_2_1.fastq.gz      /home/user/data/metagenomics/fastq/BB_2_2.fastq.gz
BB_4    /home/user/data/metagenomics/fastq/BB_4_1.fastq.gz      /home/user/data/metagenomics/fastq/BB_4_2.fastq.gz
BB_5    /home/user/data/metagenomics/fastq/BB_5_1.fastq.gz      /home/user/data/metagenomics/fastq/BB_5_2.fastq.gz
BB_6    /home/user/data/metagenomics/fastq/BB_6_1.fastq.gz      /home/user/data/metagenomics/fastq/BB_6_2.fastq.gz
BB_7    /home/user/data/metagenomics/fastq/BB_7_1.fastq.gz      /home/user/data/metagenomics/fastq/BB_7_2.fastq.gz
BB_9    /home/user/data/metagenomics/fastq/BB_9_1.fastq.gz      /home/user/data/metagenomics/fastq/BB_9_2.fastq.gz
BB_10   /home/user/data/metagenomics/fastq/BB_10_1.fastq.gz     /home/user/data/metagenomics/fastq/BB_10_2.fastq.gz
BB_11   /home/user/data/metagenomics/fastq/BB_11_1.fastq.gz     /home/user/data/metagenomics/fastq/BB_11_2.fastq.gz
BB_12   /home/user/data/metagenomics/fastq/BB_12_1.fastq.gz     /home/user/data/metagenomics/fastq/BB_12_2.fastq.gz</pre>
    Great! This should be sufficient to let QIIME2 know where the files that we want to import are.
    Now, we can import the reads with
    <pre>qiime tools import \
    --type 'SampleData[PairedEndSequencesWithQuality]' \
    --input-path fastq_abs_paths \
    --output-path fastq_imported.qza \
    --input-format PairedEndFastqManifestPhred33V2</pre>
    This hopefully finishes successfully in a few secocnds. Afterwards, you can check whether a new file was
    created with <kbd>ls</kbd> (which is slowly becoming our best friend now &ndash; right after <kbd>cd</kbd> of
    course).
    <br>
    <code>qiime</code> also requires the metadata to be in TSV (tab-separated values), whereas our file is a CSV
    (comma-separated). We can simply fix this with
    <pre>cat meta.csv | tr ',' '\t' > meta.tsv</pre>
    </p>
    <h5>Denoising with <a href="https://www.nature.com/articles/nmeth.3869">DADA2</a></h5>
    <p>
      Now that we have imported the data we can unleash the power of <code>qiime</code>! Sequence denoising (or OTU
      clustering) is the centrepiece of every 16S pipeline. We will use the <a
        href="https://www.nature.com/articles/nmeth.3869">DADA2</a> which fits a sequencing error model to the data and
      tries to merge ("denoise") sequences that differ only due to sequencing errors as opposed to actual biological
      variation.
      <pre>time qiime dada2 denoise-paired \
    --i-demultiplexed-seqs fastq_imported.qza \
    --p-trunc-len-f 190 \
    --p-trunc-len-r 190 \
    --p-n-threads 1 \
    --verbose \
    --o-table table.qza \
    --o-representative-sequences rep_seqs.qza \
    --o-denoising-stats denoising_stats.qza</pre>
      This will produce an artefact holding a list of unique sequences (<code>rep_seqs.qza</code>) as well as a table
      with the number of occurrences of each representative sequence per sample (<code>table.qza</code>).
      <a href="https://www.nature.com/articles/nmeth.3869">DADA2</a> fits the error model on all reads of a sequencing
      run simultaneously (as opposed to <a href="https://journals.asm.org/doi/10.1128/mSystems.00191-16">Deblur</a>,
      which fits it separately for each sample). It is therefore considerably slower and we will have to wait a few
      minutes
      for it to finish. In the meantime you can have a look at the later sections of the practical.
      <div class="alert alert-primary" role="alert">
        If applicable, increase the number of threads in order to speed things up.
        <br>
        Hint: You can put <code>time</code> in front of any command to see how long it took. Try it with <kbd>time sleep
          5</kbd> in a new terminal.
        <br>
        Also note that we told the program to truncate forward and reverse reads after 190&nbsp;bp due to the decrease
        in quality we saw in <code>multiqc_report.html</code>.
        As the amplicon is only expected to be ~310&nbsp;bp long, this should still give us sufficient overlap.
      </div>
    </p>
    <h5>Building a tree</h5>
    <p>
      <code>qiime</code> includes a tree-building pipeline which allows us to generate a phylogenetic tree from the
      denoised sequences with a single command. <a href="https://mafft.cbrc.jp/alignment/software/">MAFFT</a> is used
      for the alignment and multiple tree-inference methods are available (have a look at <code>qiime</code>'s <a
        href="https://docs.qiime2.org/2021.8/plugins/available/phylogeny/?highlight=phylogeny">phylogeny</a> plugin for
      details). We will use <a href="http://www.microbesonline.org/fasttree/">FastTree</a>, which is the fastest but
      also least accurate option available.
      <pre>time qiime phylogeny align-to-tree-mafft-fasttree \
    --i-sequences rep_seqs.qza \
    --o-alignment aligned_rep_seqs.qza \
    --o-masked-alignment masked_aligned_rep_seqs.qza \
    --o-tree unrooted_tree.qza \
    --o-rooted-tree rooted_tree.qza</pre>
      As you can see, this will create an alignment, mask locations that aligned badly, and then generate the tree
      (unrooted and rooted at midpoint).
      <div class="alert alert-primary" role="alert">
        We have built a new tree here because the pipeline with <a
          href="http://www.microbesonline.org/fasttree/">FastTree</a> runs quite quickly and we don't want to have to
        wait
        during the practical. However, instead of creating one from scratch, we could have also inserted our sequences
        into an existing phylogeny. <code>qiime</code> offers pre-computed trees for two popular 16S databases and an
        insertion algorithm in the <a
          href="https://docs.qiime2.org/2021.8/plugins/available/fragment-insertion/?highlight=fragment%20insertion">fragment-insertion</a>
        plugin.
        <br>
        In general, it needs to be said that phylogenomics is an incredibly deep topic. Since the tree is not
        substantial for our analysis, we can simply use
        <code>qiime</code>'s pipeline with the default parameters. However, if you ever rely on a high-quality phylogeny
        for a different project, you should definitely try to find the best approach for your data and have a close
        look at the respective literature.
      </div>
    </p>
    <h5>Estimating diversity</h5>
    <p>
      One reason for generating a phylogeny in a microbiome analysis is so that it can be used in
      phylogeny-based diversity metrics. Let's generate these now. Before we can run the corresponding command, though,
      we need to look up the lowest number of denoised reads per sample.
      <div class="alert alert-primary" role="alert">
        Large differences in sequencing depth between samples can distort the results of diversity estimates. Therefore,
        it is common practice to down-sample (in ecology-speech "<a
          href="https://en.wikipedia.org/wiki/Rarefaction_(ecology)">rarify</a>") the reads of each sample to a number
        that is equal or smaller than the number of reads in the least deeply sequenced sample. This simply means that
        we randomly select N reads from each sample where N is the smallest number of reads in any sample.
      </div>
      To check the number of denoised reads per sample, we can create a <code>qiime</code> visualisation of our counts
      table with
      <pre>time qiime feature-table summarize \
    --i-table table.qza \
    --o-visualization table.qzv</pre>
      <div class="alert alert-primary" role="alert">
        Visualisation files are produced
        by certain <code>qiime</code> commands and provide human-readable information like plots, tables, or summary
        statistics. There are several ways to view such files. The easiest one is to go to <a
          href="https://view.qiime2.org/">https://view.qiime2.org/</a> and drag & drop them into your browser window.
      </div>
      <div class="alert alert-primary" role="alert">
        Note that the tables and plots generated at <a href="https://view.qiime2.org/">https://view.qiime2.org/</a> are
        all rendered in your local browser and that nothing is uploaded to be processed on an external server, which is
        often required when working with sensitive data.
      </div>
      <div class="alert alert-primary" role="alert">
        If you are following this practical using a remote machine in the cloud, viewing these files is again going to
        be slightly more cumbersome since using a browser through SSH is going to be very slow. Instead use
        <code>scp</code> to transfer the files to your local machine or drag and drop them from MobaXterm / PuTTY (if
        you are on Windows) into the browser. If you are running macOS or Linux, you might also be able to connect your
        file browser to the remote server. Alternatively, you could also download <a
          href=https://filezilla-project.org/download.php?platform=osx">FileZilla </a>. </div> Once the visualisation
          has loaded, there should be a table looking like this: <div class="text-center"><img
            src="img/microbiome_3.png" class="img-fluid"></div>
      So, we need to rarify to a sampling depth of 24,336 reads. Let's generate the diversity metrics now (again,
      adjust the number of threads according to your setup):
      <pre>time qiime diversity core-metrics-phylogenetic \
    --i-phylogeny rooted_tree.qza \
    --i-table table.qza \
    --p-sampling-depth 24366 \
    --p-n-jobs-or-threads 1 \
    --m-metadata-file meta.tsv \
    --output-dir core-metrics-results</pre>
      This will generate a new directory <code>core-metrics-results</code> holding (based on multiple different
      diversity metrics) sample-wise diversity values ("alpha diversity"), pairwise inter-sample distance matrices
      ("beta diversity"), and visualisations of <a
        href="https://en.wikipedia.org/wiki/Multidimensional_scaling#Types">PCoA</a> plots (ending in
      <code>.qzv</code>).
      These can again be inspected with <a href="https://view.qiime2.org/">https://view.qiime2.org/</a>. For example,
      the <a href="https://en.wikipedia.org/wiki/Multidimensional_scaling#Types">PCoA</a> plot based on <a
        href="https://en.wikipedia.org/wiki/Bray%E2%80%93Curtis_dissimilarity">Bray&ndash;Curtis distance</a> with
      BV-negative samples in red and BV-positive samples in blue looks like this:
      <div class="text-center"><img src="img/microbiome_4.png" class="img-fluid" id="PCoA_plot"></div>
      <div class="alert alert-warning" role="alert">
        What does the plot tell us about our samples and the impact of BV on inter-sample diversity. Look up the term
        "Anna Karenina Principle" and what it means in terms of the microbiome. Can we say that it applies to our data?
      </div>
    </p>

    <h5>Taxonomic classification</h5>
    <p>
      We are not only interested in the ecological diversity of our samples; we also want to know which species were
      found. Again, there are multiple ways of achieving this. We will use a Naïve Bayes classifier (pre-trained on the
      <a href="https://greengenes.secondgenome.com/">Greengenes 13_8</a> database) available from the <a
        href="https://docs.qiime2.org/2021.8/data-resources/">Qiime2 website</a>. You can find it in the <code>db</code>
      directory and use it with the following command:
      <pre>time qiime feature-classifier classify-sklearn \
    --i-classifier db/gg-13-8-99-nb-classifier.qza \
    --i-reads rep_seqs.qza \
    --p-n-jobs 1 \
    --o-classification taxonomy.qza</pre>
      The produced table in <code>taxonomy.qza</code> simply links the taxonomic classifications (from phylum to species
      level) to the corresponding sequences. If you want to inspect the table, run
      <pre>time qiime metadata tabulate \
    --m-input-file taxonomy.qza \
    --o-visualization taxonomy.qzv</pre>
      and open the visualisation <code>taxonomy.qzv</code> in <a
        href="https://view.qiime2.org/">https://view.qiime2.org/</a>.

      <br><br>
      To get a more intuitive understanding of the microbial composition of our samples we can now ask
      <code>qiime</code> to plot it for us:
      <pre>time qiime taxa barplot \
    --i-table table.qza \
    --i-taxonomy taxonomy.qza \
    --m-metadata-file meta.tsv \
    --o-visualization taxa_barplot.qzv</pre>
      The resulting visualisation at species level ("Level 7") looks like this in <a
        href="https://view.qiime2.org/">https://view.qiime2.org/</a>:
      <div class="text-center"><img src="img/microbiome_5.png" class="img-fluid"></div>
      <div class="alert alert-primary" role="alert">
        You can adjust the with of the bars with the slider above the plot.
      </div>
      <div class="alert alert-warning" role="alert">
        We can see that some samples are dominated by <i>Lactobacillus iners</i> (green), whereas for others the
        situation looks very different. Double check with the metadata to find out if this is associated with
        BV-status.
      </div>
    </p>

    <h5>Testing differences in alpha diversity</h5>
    <p>
      Alpha diversity measures the general diversity of an ecosystem (i.e. a sample in our case) or a group of
      ecosystems (i.e. groups of samples like all samples with BV).
      After looking at the taxa barplot we just generated, do you think that the BV samples are statistically
      significantly more diverse than the non-BV samples? We can check if your estimate is correct by running a <a
        href="https://en.wikipedia.org/wiki/Kruskal%E2%80%93Wallis_one-way_analysis_of_variance">Kruskal&ndash;Wallis
        test</a> on the four metrics of alpha diversity that we calculated for our samples. Let's create a new
      directory to write the results into and run:
      <pre>mkdir alpha_tests</pre>
      <pre>for metric in faith_pd evenness shannon observed_features; do
    qiime diversity alpha-group-significance \
        --i-alpha-diversity core-metrics-results/${metric}_vector.qza \
        --m-metadata-file meta.tsv \
        --o-visualization alpha_tests/${metric}_group_significance.qzv
done</pre>
      This produces four more visualisations. Have a look at them. Were you right?
      <div class="alert alert-warning" role="alert">
        It looks like the different alpha diversity metrics disagree! After reading the short
        definitions of the metrics provided below, can you think of a reason for this discrepancy?
        <br>
        faith_pd: <a href="https://en.wikipedia.org/wiki/Phylogenetic_diversity">Faith's phylogenetic diversity</a> is
        defined as the sum of branch-lengths in the phylogeny between all species found in a sample (regardless of
        abundance).
        <br>
        evenness: <a href="https://en.wikipedia.org/wiki/Species_evenness">Pielou's evenness index</a> quantifies how
        differently abundant species making up an ecosystem are.
        <br>
        shannon: The <a href="https://en.wikipedia.org/wiki/Diversity_index#Shannon_index">Shannon index or Shannon
          entropy</a> quantifies how difficult it is to guess the species of a random specimen taken from the sample
        (the more species and the more equally abundant they are, the more difficult).
        <br>
        observed_features: Simply the number of unique taxa found in the sample.
      </div>
    </p>

    <h5>Testing beta diversity</h5>
    <p>As opposed to alpha diversity, which quantifies the diversity of a sample (or a group of samples) overall, beta
      diversity gives an estimate of the magnitude of differences between individual samples or groups of samples. We
      can test the
      difference between BV and non-BV samples for all beta diversity metrics with the following command:
      <pre>mkdir -p beta_tests</pre>
      <pre>for metric in bray_curtis jaccard unweighted_unifrac weighted_unifrac; do
    qiime diversity beta-group-significance \
        --i-distance-matrix core-metrics-results/${metric}_distance_matrix.qza \
        --m-metadata-file meta.tsv \
        --m-metadata-column BV \
        --o-visualization beta_tests/${metric}_significance.qzv \
        --p-permutations 9999
done</pre>
      Again, we got a visualisation for each metric. Have a look at them in <a
        href="https://view.qiime2.org/">https://view.qiime2.org/</a>. What do you find? Are all of them in
      agreement this time?
      <br><br>
      This concludes today's practical. If you are interested in differentially abundant taxa between the BV and non-BV samples, have a look at <code>qiime</code>'s' <a href="https://docs.qiime2.org/2021.8/plugins/available/composition/ancom/?highlight=ancom">ANCOM function</a>.
    </p>


    <br><br>
    <p><b>Acknowledgements: </b>Many thanks to Dr. Suzanna Francis for providing the data.</p>

  </div>




  <!-- Optional JavaScript -->
  <!-- jQuery first, then Popper.js, then Bootstrap JS -->
  <script src="jquery-3.3.1.slim.min.js" crossorigin="anonymous"></script>
  <script src="popper.min.js" crossorigin="anonymous"></script>
  <script src="bootstrap.min.js" crossorigin="anonymous"></script>
  <script src="d3.v3.min.js"></script>
  <script src="stack.v1.min.js"></script>

</body>

</html>